builtin/plugin/v5/backend.go

```diff
@@ -0,0 +1,147 @@
+package plugin
+
+import (
+	"context"
+	"net/rpc"
+	"sync"
+
+	uuid "github.com/hashicorp/go-uuid"
+	"github.com/hashicorp/vault/sdk/helper/consts"
+	"github.com/hashicorp/vault/sdk/logical"
+	bplugin "github.com/hashicorp/vault/sdk/plugin"
+	v5 "github.com/hashicorp/vault/sdk/plugin/v5"
+)
+
+// Backend returns an instance of the backend, either as a plugin if external
+// or as a concrete implementation if builtin, casted as logical.Backend.
+func Backend(ctx context.Context, conf *logical.BackendConfig) (logical.Backend, error) {
+	var b backend
+	name := conf.Config["plugin_name"]
+	pluginType, err := consts.ParsePluginType(conf.Config["plugin_type"])
+	if err != nil {
+		return nil, err
+	}
+
+	sys := conf.System
+
+	raw, err := v5.NewBackend(ctx, name, pluginType, sys, conf)
+	if err != nil {
+		return nil, err
+	}
+	b.Backend = raw
+	b.config = conf
+
+	return &b, nil
+}
+
+// backend is a thin wrapper around plugin.BackendPluginClient
+type backend struct {
+	logical.Backend
+	mu sync.RWMutex
+
+	config *logical.BackendConfig
+
+	// Used to detect if we already reloaded
+	canary string
+}
+
+func (b *backend) reloadBackend(ctx context.Context) error {
+	pluginName := b.config.Config["plugin_name"]
+	pluginType, err := consts.ParsePluginType(b.config.Config["plugin_type"])
+	if err != nil {
+		return err
+	}
+
+	b.Logger().Debug("plugin: reloading plugin backend", "plugin", pluginName)
+
+	// Ensure proper cleanup of the backend
+	// Pass a context value so that the plugin client will call the appropriate
+	// cleanup method for reloading
+	reloadCtx := context.WithValue(ctx, v5.ContextKeyPluginReload, "reload")
+	b.Backend.Cleanup(reloadCtx)
+
+	nb, err := v5.NewBackend(ctx, pluginName, pluginType, b.config.System, b.config)
+	if err != nil {
+		return err
+	}
+	err = nb.Setup(ctx, b.config)
+	if err != nil {
+		return err
+	}
+	b.Backend = nb
```

Comment:
🤔 Do we need to call initialize here too?

Reply:
No, that gets called in the respective [auth.go](https://github.com/hashicorp/vault/blob/a222dbfd2954bf2f41e27fb96f4d9be2347504f7/vault/auth.go#L219) and mount.go

Reply:
But in the case of plugin crash and is restrated i think we'd need to re-initialize, right?